In 0.16.0, cannot build tests when using system c4core

I appreciate the new `RYML_SYSTEM_C4CORE` option in `rapydyaml` 0.16.0. I already build `rapidyaml` with the system `c4core` in Fedora, and this option will make that more straightforward. However, I would like to test the package after I build it, and I’m unable to combine `-DRYML_SYSTEM_C4CORE:BOOL=ON` with `-DRYML_BUILD_TESTS:BOOL=ON`. The following is from a git checkout,  with a `c4core-devel` 0.6.0 package installed system-wide.

```
$ cmake -S. -Bbuild -DRYML_BUILD_TESTS:BOOL=ON -DRYML_SYSTEM_C4CORE:BOOL=ON
-- The CXX compiler identification is GNU 16.1.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- ryml: using C++ standard: C++11
-- ryml: setting C++ standard: 11
-- ryml: setting C++ standard required: ON
-- ryml: downloading c4fs: searching cached project...
-- ryml: downloading c4fs: not in cache...
-- ryml: downloading remote project: c4fs -> "/home/ben/src/forks/rapidyaml/build/subprojects/c4fs/src" (dir=/home/ben/src/forks/rapidyaml/build/subprojects/c4fs)...
CMake Deprecation Warning at CMakeLists.txt:2 (cmake_minimum_required):
  Compatibility with CMake < 3.10 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value.  Or, use the <min>...<max> syntax
  to tell CMake that the project requires at least <min> but has been updated
  to work with policies introduced by <max> or earlier.


-- Configuring done (0.0s)
-- Generating done (0.0s)
-- Build files have been written to: /home/ben/src/forks/rapidyaml/build/subprojects/c4fs/dl
[ 11%] Creating directories for 'c4fs-dl'
[ 22%] Performing download step (git clone) for 'c4fs-dl'
Cloning into 'src'...
Already on 'master'
Your branch is up to date with 'origin/master'.
Submodule 'ext/c4core' (https://github.com/biojppm/c4core) registered for path 'ext/c4core'
Cloning into '/home/ben/src/forks/rapidyaml/build/subprojects/c4fs/src/ext/c4core'...
Submodule path 'ext/c4core': checked out '7f3863e15ffcb2eef386d899c9c63b4db9418979'
Submodule 'cmake' (https://github.com/biojppm/cmake) registered for path 'ext/c4core/cmake'
Submodule 'extern/debugbreak' (https://github.com/biojppm/debugbreak) registered for path 'ext/c4core/src/c4/ext/debugbreak'
Submodule 'src/c4/ext/fast_float' (https://github.com/fastfloat/fast_float) registered for path 'ext/c4core/src/c4/ext/fast_float'
Cloning into '/home/ben/src/forks/rapidyaml/build/subprojects/c4fs/src/ext/c4core/cmake'...
Cloning into '/home/ben/src/forks/rapidyaml/build/subprojects/c4fs/src/ext/c4core/src/c4/ext/debugbreak'...
Cloning into '/home/ben/src/forks/rapidyaml/build/subprojects/c4fs/src/ext/c4core/src/c4/ext/fast_float'...
Submodule path 'ext/c4core/cmake': checked out 'aee129e31f4a7c7a200fb63559ab66da4101c642'
Submodule path 'ext/c4core/src/c4/ext/debugbreak': checked out '328e4abca3384cbd0a69e70f263cc7b2794bff09'
Submodule path 'ext/c4core/src/c4/ext/fast_float': checked out 'eb9ab42c0aa86d3cc41871ba7306527d95077b3b'
[ 33%] Performing update step for 'c4fs-dl'
-- Fetching latest from the remote origin
[ 44%] No patch step for 'c4fs-dl'
[ 55%] No configure step for 'c4fs-dl'
[ 66%] No build step for 'c4fs-dl'
[ 77%] No install step for 'c4fs-dl'
[ 88%] No test step for 'c4fs-dl'
[100%] Completed 'c4fs-dl'
[100%] Built target c4fs-dl
CMake Error at ext/testbm.cmake:20 (ryml_add_c4core_dev_to_target):
  Unknown CMake command "ryml_add_c4core_dev_to_target".
Call Stack (most recent call first):
  CMakeLists.txt:194 (include)


-- Configuring incomplete, errors occurred!
```

The issue is that tests require `ext/testbm.cmake`, which calls `ryml_add_c4core_dev_to_target`:

https://github.com/biojppm/rapidyaml/blob/f8ac8dd50f4f7916579d55a05ebf9c6488e52670/ext/testbm.cmake#L20
https://github.com/biojppm/rapidyaml/blob/f8ac8dd50f4f7916579d55a05ebf9c6488e52670/ext/testbm.cmake#L45

But this is defined in `ext/c4core.cmake`,

https://github.com/biojppm/rapidyaml/blob/f8ac8dd50f4f7916579d55a05ebf9c6488e52670/ext/c4core.cmake#L10

and that is only included in the top-level `CMakeLists.txt` when we are doing a standalone build of `rapidyaml` and are *not* using the system `c4core`:

https://github.com/biojppm/rapidyaml/blob/f8ac8dd50f4f7916579d55a05ebf9c6488e52670/CMakeLists.txt#L99-L107

I could probably figure out a workaround for this, or perhaps even a proper fix, but I suspect that you will immediately know what to do about this.

--- a/.github/workflows/install.yml
+++ b/.github/workflows/install.yml
@@ -213,3 +213,67 @@ jobs:
       run: |
         cd $BDIR/test
         ctest -C ${{matrix.bt}} --output-on-failure
+  sysc4core:
+    name: sysc4core/${{matrix.name}}
+    continue-on-error: false
+    runs-on: ${{matrix.os}}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+        - name: gcc
+          os: ubuntu-24.04
+          gen: -DCMAKE_CXX_COMPILER=g++
+          bt: Debug
+          rel: lib/cmake/c4core
+        - name: clang
+          os: ubuntu-24.04
+          gen: -DCMAKE_CXX_COMPILER=clang++
+          bt: Debug
+          rel: lib/cmake/c4core
+        - name: xcode
+          os: macos-15-intel
+          gen: -G Xcode -DCMAKE_OSX_ARCHITECTURES=x86_64
+          bt: Debug
+          rel: lib/cmake/c4core
+        - name: vs2026
+          os: windows-latest
+          gen: -G 'Visual Studio 18 2026' -A x64
+          bt: Debug
+          rel: cmake
+    steps:
+    - name: checkout
+      uses: actions/checkout@v6
+    - name: checkout c4core
+      run: make -C ext c4core
+    - name: config c4core
+      run: |
+        cmake -S ext/c4core -B ext/c4core/build \
+           -D CMAKE_INSTALL_PREFIX=`pwd`/ext/c4core/install \
+           ${{matrix.gen}}
+    - name: build c4core
+      run: |
+        cmake --build ext/c4core/build --target c4core --parallel
+    - name: install c4core
+      run: |
+        cmake --build ext/c4core/build --target install
+    - name: '--------------------------------------'
+      run: echo
+    - name: configure ryml
+      run: |
+        cmake -S . -B build \
+          -D c4core_DIR=`pwd`/ext/c4core/install/${{matrix.rel}} \
+          -D RYML_SYSTEM_C4CORE=ON \
+          -D RYML_BUILD_TESTS=ON \
+          ${{matrix.gen}} \
+          -D CMAKE_BUILD_TYPE=${{matrix.bt}}
+    - name: build ryml
+      run: |
+        cmake --build build --target ryml --parallel --config ${{matrix.bt}}
+    - name: build tests
+      run: |
+        cmake --build build --target ryml-test-build --parallel --config ${{matrix.bt}}
+    - name: run tests
+      run: |
+        export CTEST_PARALLEL_LEVEL=`nproc`
+        cmake --build build --target ryml-test-run --config ${{matrix.bt}}

--- a/.github/workflows/install.ys
+++ b/.github/workflows/install.ys
@@ -11,6 +11,8 @@ env:
   NUM_JOBS_BUILD: # 4
 
 jobs:
+
+  #---------------------------
   install_tests:
     if: |
       (!contains(github.event.head_commit.message, 'skip all')) ||
@@ -19,32 +21,26 @@ jobs:
     name: ${{matrix.name}}/${{matrix.bt}}
     continue-on-error: true
     runs-on: ${{matrix.os}}
-
     strategy:
       fail-fast: false
       matrix:
         include:
-
           - {name: find_package/linux,        sdir: test/test_install,      os: ubuntu-24.04,   cxx: g++-16,     gen: -DCMAKE_CXX_COMPILER=g++-16,               tgt: all,       bt: Debug,   vars: -Dryml_DIR=$GITHUB_WORKSPACE/$PDIR/lib/cmake/ryml -DRYML_TEST_INSTALL_PACKAGE_MODE=ON}
           - {name: find_package/linux/libcxx, sdir: test/test_install,      os: ubuntu-24.04,   cxx: clang++-18, gen: -DCMAKE_CXX_COMPILER=clang++-18,           tgt: all,       bt: Debug,   vars: -Dryml_DIR=$GITHUB_WORKSPACE/$PDIR/lib/cmake/ryml -DRYML_TEST_INSTALL_PACKAGE_MODE=ON, commonvars: -DRYML_USE_LIBCXX=ON}
           - {name: find_package/macos,        sdir: test/test_install,      os: macos-15-intel, cxx: xcode,      gen: -G Xcode -DCMAKE_OSX_ARCHITECTURES=x86_64, tgt: ALL_BUILD, bt: Debug,   vars: -Dryml_DIR=$GITHUB_WORKSPACE/$PDIR/lib/cmake/ryml -DRYML_TEST_INSTALL_PACKAGE_MODE=ON}
           - {name: find_package/win,          sdir: test/test_install,      os: windows-latest, cxx: vs2026,     gen: -G 'Visual Studio 18 2026' -A x64,         tgt: ALL_BUILD, bt: Debug,   vars: -Dryml_DIR=$GITHUB_WORKSPACE/$PDIR/cmake -DRYML_TEST_INSTALL_PACKAGE_MODE=ON}
-
           - {name: find_library/linux,        sdir: test/test_install,      os: ubuntu-24.04,   cxx: g++-16,     gen: -DCMAKE_CXX_COMPILER=g++-16,               tgt: all,       bt: Debug,   vars: -DCMAKE_PREFIX_PATH=$GITHUB_WORKSPACE/$PDIR -DRYML_TEST_INSTALL_PACKAGE_MODE=OFF}
           - {name: find_library/linux/libcxx, sdir: test/test_install,      os: ubuntu-24.04,   cxx: clang++-18, gen: -DCMAKE_CXX_COMPILER=clang++-18,           tgt: all,       bt: Debug,   vars: -DCMAKE_PREFIX_PATH=$GITHUB_WORKSPACE/$PDIR -DRYML_TEST_INSTALL_PACKAGE_MODE=OFF, commonvars: -DRYML_USE_LIBCXX=ON}
           - {name: find_library/macos,        sdir: test/test_install,      os: macos-15-intel, cxx: xcode,      gen: -G Xcode -DCMAKE_OSX_ARCHITECTURES=x86_64, tgt: ALL_BUILD, bt: Debug,   vars: -DCMAKE_PREFIX_PATH=$GITHUB_WORKSPACE/$PDIR -DRYML_TEST_INSTALL_PACKAGE_MODE=OFF}
           - {name: find_library/win,          sdir: test/test_install,      os: windows-latest, cxx: vs2026,     gen: -G 'Visual Studio 18 2026' -A x64,         tgt: ALL_BUILD, bt: Debug,   vars: -DCMAKE_PREFIX_PATH=$GITHUB_WORKSPACE/$PDIR -DRYML_TEST_INSTALL_PACKAGE_MODE=OFF}
-
           - {name: singleheader/linux,        sdir: test/test_singleheader, os: ubuntu-24.04,   cxx: g++-16,     gen: -DCMAKE_CXX_COMPILER=g++-16,               tgt: all,       bt: Debug}
           - {name: singleheader/linux/libcxx, sdir: test/test_singleheader, os: ubuntu-24.04,   cxx: clang++-18, gen: -DCMAKE_CXX_COMPILER=clang++-18,           tgt: all,       bt: Debug, commonvars: -DRYML_USE_LIBCXX=ON}
           - {name: singleheader/macos,        sdir: test/test_singleheader, os: macos-15-intel, cxx: xcode,      gen: -G Xcode -DCMAKE_OSX_ARCHITECTURES=x86_64, tgt: ALL_BUILD, bt: Debug}
           - {name: singleheader/win,          sdir: test/test_singleheader, os: windows-latest, cxx: vs2026,     gen: -G 'Visual Studio 18 2026' -A x64,         tgt: ALL_BUILD, bt: Debug}
-
           - {name: singlesource/linux,        sdir: test/test_singlesource, os: ubuntu-24.04,   cxx: g++-16,     gen: -DCMAKE_CXX_COMPILER=g++-16,               tgt: all,       bt: Debug}
           - {name: singlesource/linux/libcxx, sdir: test/test_singlesource, os: ubuntu-24.04,   cxx: clang++-18, gen: -DCMAKE_CXX_COMPILER=clang++-18,           tgt: all,       bt: Debug, commonvars: -DRYML_USE_LIBCXX=ON}
           - {name: singlesource/macos,        sdir: test/test_singlesource, os: macos-15-intel, cxx: xcode,      gen: -G Xcode -DCMAKE_OSX_ARCHITECTURES=x86_64, tgt: ALL_BUILD, bt: Debug}
           - {name: singlesource/win,          sdir: test/test_singlesource, os: windows-latest, cxx: vs2026,     gen: -G 'Visual Studio 18 2026' -A x64,         tgt: ALL_BUILD, bt: Debug}
-
     env:
       CXX_: ${{matrix.cxx}}
       BT:   ${{matrix.bt}}
@@ -52,7 +48,6 @@ jobs:
       BDIR: build/${{matrix.name}}-${{matrix.bt}}
       IDIR: install/${{matrix.name}}-${{matrix.bt}}
       PDIR: prefix/${{matrix.name}}-${{matrix.bt}}
-
     steps:
     - name: checkout
       uses: actions/checkout@v6
@@ -101,3 +96,55 @@ jobs:
       run: |
         cd $BDIR/test
         ctest -C ${{matrix.bt}} --output-on-failure
+
+
+  #---------------------------
+  sysc4core:
+    name: sysc4core/${{matrix.name}}
+    continue-on-error: false
+    runs-on: ${{matrix.os}}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - {name: gcc,    os: ubuntu-24.04,   gen: -DCMAKE_CXX_COMPILER=g++,                  bt: Debug, rel: lib/cmake/c4core }
+          - {name: clang,  os: ubuntu-24.04,   gen: -DCMAKE_CXX_COMPILER=clang++,              bt: Debug, rel: lib/cmake/c4core }
+          - {name: xcode,  os: macos-15-intel, gen: -G Xcode -DCMAKE_OSX_ARCHITECTURES=x86_64, bt: Debug, rel: lib/cmake/c4core }
+          - {name: vs2026, os: windows-latest, gen: -G 'Visual Studio 18 2026' -A x64,         bt: Debug, rel: cmake }
+    steps:
+    - name: checkout
+      uses: actions/checkout@v6
+    - name: checkout c4core
+      run: make -C ext c4core
+    - name: config c4core
+      run: |
+        cmake -S ext/c4core -B ext/c4core/build \
+           -D CMAKE_INSTALL_PREFIX=`pwd`/ext/c4core/install \
+           ${{matrix.gen}}
+    - name: build c4core
+      run: |
+        cmake --build ext/c4core/build --target c4core --parallel
+    - name: install c4core
+      run: |
+        cmake --build ext/c4core/build --target install
+    - name: "--------------------------------------"
+      run: echo
+    - name: configure ryml
+      run: |
+        cmake -S . -B build \
+          -D c4core_DIR=`pwd`/ext/c4core/install/${{matrix.rel}} \
+          -D RYML_SYSTEM_C4CORE=ON \
+          -D RYML_BUILD_TESTS=ON \
+          ${{matrix.gen}} \
+          -D CMAKE_BUILD_TYPE=${{matrix.bt}}
+    - name: build ryml
+      run: |
+        cmake --build build --target ryml --parallel --config ${{matrix.bt}}
+    - name: build tests
+      run: |
+        cmake --build build --target ryml-test-build --parallel --config ${{matrix.bt}}
+    - name: run tests
+      run: |
+        export CTEST_PARALLEL_LEVEL=`nproc`
+        cmake --build build --target ryml-test-run --config ${{matrix.bt}}
+

--- a/changelog/current.md
+++ b/changelog/current.md
@@ -1,0 +1,1 @@
+- [PR#653](https://github.com/biojppm/rapidyaml/pull/653) cmake: fix tests with RYML_SYSTEM_C4CORE (fixes [#651](https://github.com/biojppm/rapidyaml/issues/651))

--- a/ext/testbm.cmake
+++ b/ext/testbm.cmake
@@ -1,5 +1,21 @@
 # these are used both for testing and benchmarking
 
+function(ryml_testbm_link_with_c4core target)
+    if(RYML_SYSTEM_C4CORE)
+        target_link_libraries(${target} PUBLIC c4core::c4core)
+    elseif(NOT RYML_STANDALONE)
+        target_link_libraries(${target} PUBLIC c4core)
+    else() # RYML_STANDALONE
+        target_link_libraries(${target} PUBLIC ryml) # link with ryml to get c4core
+        ryml_add_c4core_dev_to_target(${target})
+        target_include_directories(${target} PUBLIC
+            ${CMAKE_CURRENT_LIST_DIR}/c4core.src
+            ${CMAKE_CURRENT_LIST_DIR}/c4core.dev
+        )
+    endif()
+endfunction()
+
+
 set(C4FS_DIR ${CMAKE_CURRENT_BINARY_DIR}/subprojects/c4fs)
 c4_download_remote_proj(c4fs C4FS_DIR
     GIT_REPOSITORY https://github.com/biojppm/c4fs
@@ -9,15 +25,11 @@ c4_add_library(c4fs LIBRARY_TYPE STATIC
         ${C4FS_DIR}/src/c4/fs/fs.hpp
         ${C4FS_DIR}/src/c4/fs/fs.cpp
     INC_DIRS
-        ${CMAKE_CURRENT_LIST_DIR}/c4core.src
-        ${CMAKE_CURRENT_LIST_DIR}/c4core.dev
         ${C4FS_DIR}/src
-    LIBS
-        ryml # link with ryml to get c4core
     FOLDER
         ext
 )
-ryml_add_c4core_dev_to_target(c4fs)
+ryml_testbm_link_with_c4core(c4fs)
 if(WIN32 AND BUILD_SHARED_LIBS)
     target_compile_definitions(c4fs PUBLIC C4FS_SHARED
         $<BUILD_INTERFACE:C4FS_EXPORTS>
@@ -34,15 +46,11 @@ c4_add_library(c4log
         ${C4LOG_DIR}/src/c4/log/log.hpp
         ${C4LOG_DIR}/src/c4/log/log.cpp
     INC_DIRS
-        ${CMAKE_CURRENT_LIST_DIR}/c4core.src
-        ${CMAKE_CURRENT_LIST_DIR}/c4core.dev
         ${C4LOG_DIR}/src
-    LIBS
-        ryml # link with ryml to get c4core
     FOLDER
         ext
 )
-ryml_add_c4core_dev_to_target(c4log)
+ryml_testbm_link_with_c4core(c4log)
 if(WIN32 AND BUILD_SHARED_LIBS)
     target_compile_definitions(c4log PUBLIC C4LOG_SHARED
         $<BUILD_INTERFACE:C4LOG_EXPORTS>

--- a/samples/custom_c4core/run.sh
+++ b/samples/custom_c4core/run.sh
@@ -17,7 +17,7 @@ if [ "$c4core_tag" == "." ] || [ -z "$c4core_tag" ] ; then
     c4core_tag=$(cat ../../ext/c4core.mk \
                  | grep C4CORE_TAG \
                  | grep = \
-                 | sed 's/.*=\s*\([0-9a-fA-F]*\)/\1/g')
+                 | sed 's/.*=\s*\(.*\)/\1/g')
 fi
 
 # make sure to run from where this file is

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -114,7 +114,7 @@ c4_add_library(ryml-_testlib LIBRARY_TYPE STATIC
     LIBS ryml c4fs gtest
     FOLDER test)
 ryml_setup_test_target_props(ryml-_testlib)
-ryml_add_c4core_dev_to_target(ryml-_testlib)
+ryml_testbm_link_with_c4core(ryml-_testlib)
 c4_add_library(ryml-_testmain LIBRARY_TYPE OBJECT
     SOURCES test_lib/test_main.cpp test_lib/test_main.hpp
     LIBS ryml
@@ -128,7 +128,7 @@ function(ryml_add_test test_name)
         SOURCES test_${test_name}.cpp
         LIBS ryml-_testlib ryml-_testmain
         FOLDER test)
-    ryml_add_c4core_dev_to_target(${t})
+    ryml_testbm_link_with_c4core(${t})
     if(RYML_SAVE_TEST_YAML)
         target_sources(${t} PRIVATE test_lib/test_save.cpp)
     endif()
@@ -147,7 +147,7 @@ c4_add_library(ryml-_testgroup LIBRARY_TYPE STATIC
     LIBS ryml ryml-_testlib c4fs
     FOLDER test)
 ryml_setup_test_target_props(ryml-_testgroup)
-ryml_add_c4core_dev_to_target(ryml-_testgroup)
+ryml_testbm_link_with_c4core(ryml-_testgroup)
 function(ryml_add_test_case_group name)
     ryml_add_test(${name} ${ARGN})
     target_link_libraries(ryml-test-${name} PUBLIC ryml-_testgroup)
@@ -399,7 +399,7 @@ if(RYML_TEST_SUITE)
         LIBS ryml-_testlib ryml-_testmain c4log
         INC_DIRS ${CMAKE_CURRENT_LIST_DIR}/../src_extra
         FOLDER test)
-    ryml_add_c4core_dev_to_target(ryml-test-suite)
+    ryml_testbm_link_with_c4core(ryml-test-suite)
     add_dependencies(ryml-test-build ryml-test-suite)
 
     ryml_get_target_exe(ryml-test-suite tgt)

--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -6,5 +6,5 @@ c4_add_executable(ryml-parse-emit
     INC_DIRS ../test
     LIBS ryml c4fs
     FOLDER tools)
-ryml_add_c4core_dev_to_target(ryml-parse-emit)
+ryml_testbm_link_with_c4core(ryml-parse-emit)
 ryml_maybe_dump_test_yaml(ryml-parse-emit)

